webodf/lib/gui/TrivialUndoManager.js

```diff
@@ -37,32 +37,124 @@ gui.TrivialUndoManager = function TrivialUndoManager(defaultRules) {
         domUtils = core.DomUtils,
         /**@type{?Element}*/
         initialDoc,
-        /**@type{!Array.<!ops.Operation>}*/
-        initialState = [],
+        /**@type{!StateTransition}*/
+        initialStateTransition,
         playFunc,
         /**@type{!ops.Document}*/
         document,
-        /**@type{!Array.<!ops.Operation>}*/
-        currentUndoState = [],
-        /**@type{!Array.<!Array.<!ops.Operation>>}*/
-        undoStates = [],
-        /**@type{!Array.<!Array.<!ops.Operation>>}*/
-        redoStates = [],
+        /**@type {!StateId}*/
+        unmodifiedStateId,
+        /**@type{!StateTransition}*/
+        currentUndoStateTransition,
+        /**@type{!Array.<!StateTransition>}*/
+        undoStateTransitions = [],
+        /**@type{!Array.<!StateTransition>}*/
+        redoStateTransitions = [],
         eventNotifier = new core.EventNotifier([
             gui.UndoManager.signalUndoStackChanged,
             gui.UndoManager.signalUndoStateCreated,
             gui.UndoManager.signalUndoStateModified,
+            gui.UndoManager.signalDocumentModifiedChanged,
             gui.TrivialUndoManager.signalDocumentRootReplaced
         ]),
         undoRules = defaultRules || new gui.UndoStateRules(),
-        isExecutingOps = false;
+        isExecutingOps = false,
+        /**
+        * Base for generating unique state ids
+        * @type {!number}
+        */
+        stateIdBase = 0;
+
+    /**
+     * @constructor
+     * @param {!number=} mainId
+     * @param {!number=} subId
+     */
+    function StateId(mainId, subId) {
```

Comment:
Hm... torn between lazyness and supporting easy-to-read-code... ;) But then I always demand the latter, so seems I have no real choice.

((For the runtime efficiency, that might be still solvable one day by creating another context, e.g. StateTransitionFactory, where such an object gets the `undoRules` object passed, has a method `createStateTransition` and `StateTransition` as internally defined classes. But well, no real idea about JS runtime and what JS code results in what, so leaving for when really a measured issue :question: :arrow_right: :zzz: ))